activerecord/lib/active_record/insert_all.rb

```diff
@@ -105,13 +108,14 @@ def values_list
 
           values_list = insert_all.inserts.map do |attributes|
             attributes = attributes.stringify_keys
+            attributes_keys_with_defaults = scope_for_create.keys | attributes.keys | defaults.keys
 
-            unless attributes.keys.to_set == keys
+            unless attributes_keys_with_defaults.to_set == keys
               raise ArgumentError, "All objects being inserted must have the same keys"
             end
 
             keys.map do |key|
-              bind = Relation::QueryAttribute.new(key, attributes[key], types[key])
+              bind = Relation::QueryAttribute.new(key, scope_for_create[key] || attributes[key] || defaults[key], types[key])
               connection.with_yaml_fallback(bind.value_for_database)
```

Comment:
@kamipo would it make sense to extract the below from here into something like `bind_value_for_database`?

```ruby
bind = Relation::QueryAttribute.new(key, attributes[key], types[key])
connection.with_yaml_fallback(bind.value_for_database)
```

InsertAll has a fairly intimate relationship with connection and also sort of Relation::QueryAttribute. I guess the tough part is `types[key]` because we might not want to run `lookup_cast_type_from_column` on the M*N loop.